jax_distribution/systems/ippo_ff_anakin.py

```diff
@@ -0,0 +1,475 @@
+        # timestep.extras = {}
+        state = LogEnvState(state, 0.0, 0, 0.0, 0)
+        return state, timestep
+
+    def step(
+        self,
+        state: State,
+        action: jnp.ndarray,
+    ) -> Tuple[State, TimeStep]:
+
+        env_state, timestep = self._env.step(state.env_state, action)
+
+        new_episode_return = state.episode_returns + jnp.mean(timestep.reward)
+        new_episode_length = state.episode_lengths + 1
+        state = LogEnvState(
+            env_state=env_state,
+            episode_returns=new_episode_return * (1 - timestep.last()),
+            episode_lengths=new_episode_length * (1 - timestep.last()),
+            returned_episode_returns=state.returned_episode_returns
+            * (1 - timestep.last())
+            + new_episode_return * timestep.last(),
+            returned_episode_lengths=state.returned_episode_lengths
+            * (1 - timestep.last())
+            + new_episode_length * timestep.last(),
+        )
+        return state, timestep
+
+
+class Transition(NamedTuple):
+    done: jnp.ndarray
+    action: jnp.ndarray
+    value: jnp.ndarray
+    reward: jnp.ndarray
+    log_prob: jnp.ndarray
+    observation: jnp.ndarray
+    info: jnp.ndarray
+
+
+class ActorCritic(nn.Module):
+    action_dim: Sequence[int]
+    activation: str = "tanh"
+    env_name: str = "RobotWarehouse-v0"
+
+    @nn.compact
+    def __call__(self, observation):
+        if "MultiCVRP" in self.env_name:
+            x = observation.vehicles.coordinates  # .shape
+        elif "RobotWarehouse" in self.env_name:
+            x = observation.agents_view
+
+        if self.activation == "relu":
+            activation = nn.relu
+        else:
+            activation = nn.tanh
+
+        actor_mean = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(x)
+
+        actor_mean = activation(actor_mean)
+        actor_mean = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(actor_mean)
+        actor_mean = activation(actor_mean)
+
+        actor_mean = nn.Dense(
+            self.action_dim, kernel_init=orthogonal(0.01), bias_init=constant(0.0)
+        )(actor_mean)
+
+        masked_logits = jnp.where(
+            observation.action_mask,
+            actor_mean,
+            jnp.finfo(jnp.float32).min,
+        )
+
+        pi = distrax.Categorical(logits=masked_logits)
+
+        critic = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(x)
+        critic = activation(critic)
+        critic = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(critic)
+        critic = activation(critic)
+        critic = nn.Dense(1, kernel_init=orthogonal(1.0), bias_init=constant(0.0))(
+            critic
+        )
+
+        return pi, jnp.squeeze(critic, axis=-1)
+
+
+def get_learner_fn(env, forward_pass, opt_update, config):
```

Comment:
I'm happy to help with the type hints here.

Reply:
Done thanks @RuanJohn